### PR TITLE
XCM simulator

### PR DESCRIPTION
**Pull request by shaunxw**
Compared to the original [xcm-simulator](https://github.com/shaunxw/xcm-simulator), all cumulus dependencies are removed. The new impl uses `XcmpMessageHandler` and `DmpMessageHandler` abstractions, instead of directly calling cumulus pallets to handle xcmp and dmp. Users need to pass these implementations when declaring test parachain.

In order to make the new impl work, the following definitions are moved to polkadot from cumulus(will need a cumulus companion PR):
- `XcmpMessageHandler` trait
- `DmpMessageHandler` trait
- `XcmpMessageFormat` enum

### CODE REVIEW

File: xcm/xcm-simulator/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+			let encoded = VersionedXcm::<()>::from(message).encode();
+			outbound.extend_from_slice(&encoded[..]);
+			outbound
+		},
+	}
+}
+
+#[macro_export]
+macro_rules! decl_test_relay_chain {
```
Comment (apopiak) [inline xcm/xcm-simulator/src/lib.rs:65]: It would be great if you could add some doc comments for these macros.

**Review by apopiak — APPROVED** [review]
Awesome contribution to allow for integration testing :+1:

**Review by shawntabrizi — APPROVED** [review]

File: xcm/xcm-simulator/src/lib.rs (hunk 1 above)
Comment (shawntabrizi) [reply xcm/xcm-simulator/src/lib.rs:65]: lets follow up on that :)

https://github.com/paritytech/polkadot/issues/3542

**Comment by shawntabrizi** [comment]
bot merge

**Comment by jhonalino** [comment]
such a beautiful piece of code

**Pull request merged**